source/client.c

```diff
@@ -1788,3 +1779,150 @@ int aws_mqtt_client_connection_ping(struct aws_mqtt_client_connection *connectio
+            return AWS_MQTT_CLIENT_REQUEST_ERROR;
+        }
+
+        const size_t num_topics = aws_array_list_length(&task_arg->topics);
+        if (num_topics <= 0) {
+            aws_raise_error(AWS_ERROR_MQTT_INVALID_TOPIC);
+            return AWS_MQTT_CLIENT_REQUEST_ERROR;
+        }
+
+        for (size_t i = 0; i < num_topics; ++i) {
+
+            struct aws_mqtt_topic_subscription *topic = NULL;
+            aws_array_list_get_at_ptr(&task_arg->topics, (void **)&topic, i);
+            AWS_ASSUME(topic); /* We know we're within bounds */
+
+            if (aws_mqtt_packet_subscribe_add_topic(&task_arg->subscribe, topic->topic, topic->qos)) {
+                goto handle_error;
+            }
+        }
+    }
+
+    message = mqtt_get_message_for_packet(task_arg->connection, &task_arg->subscribe.fixed_header);
+    if (!message) {
+
+        goto handle_error;
+    }
+
+    if (aws_mqtt_packet_subscribe_encode(&message->message_data, &task_arg->subscribe)) {
+
+        goto handle_error;
+    }
+
+    /* No need to handle this error, if the send fails, it'll just retry */
+    aws_channel_slot_send_message(task_arg->connection->slot, message, AWS_CHANNEL_DIR_WRITE);
+
+    return AWS_MQTT_CLIENT_REQUEST_ONGOING;
+
+handle_error:
+
+    if (message) {
+        aws_mem_release(message->allocator, message);
+    }
+
+    return AWS_MQTT_CLIENT_REQUEST_ERROR;
+}
+
+static void s_resubscribe_complete(
+    struct aws_mqtt_client_connection *connection,
+    uint16_t packet_id,
+    int error_code,
+    void *userdata) {
+
+    struct resubscribe_task_arg *task_arg = userdata;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_MQTT_CLIENT,
+        "id=%p: Resubscribe %" PRIu16 " completed with error_code %d",
+        (void *)connection,
+        packet_id,
+        error_code);
+
+    aws_array_list_clean_up(&task_arg->topics);
+    aws_mqtt_packet_subscribe_clean_up(&task_arg->subscribe);
+    aws_mem_release(task_arg->connection->allocator, task_arg);
+}
+
+void mqtt_resubscribe_existing_topics(struct aws_mqtt_client_connection *connection) {
+    size_t sub_count = aws_mqtt_topic_tree_get_sub_count(&connection->subscriptions);
+    static const size_t sub_size = sizeof(struct aws_mqtt_topic_subscription);
+
+    if (sub_count == 0) {
+        return;
+    }
+
+    struct resubscribe_task_arg *task_arg = NULL;
+    void *buffer = NULL;
+    aws_mem_acquire_many(
+        connection->allocator, 2, &task_arg, sizeof(struct resubscribe_task_arg), &buffer, sub_count * sub_size);
+
+    if (!task_arg) {
+        AWS_LOGF_FATAL(
+            AWS_LS_MQTT_CLIENT, "id=%p: failed to allocate storage for resubscribe arguments", (void *)connection);
+        return;
+    }
+
+    AWS_ZERO_STRUCT(*task_arg);
+    task_arg->connection = connection;
+    aws_array_list_init_static(&task_arg->topics, buffer, sub_count, sub_size);
+    aws_mqtt_topic_tree_iterate(&connection->subscriptions, s_reconnect_resub_iterator, task_arg);
```

Comment:
is there anything that might go wrong in the iterator function, that we'd want to check for after iterating everything?

Reply:
Nope! It's just pushing into an array list of a known good size.